### PR TITLE
upgrade velocypack to latest version

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Upgrade velocypack version to latest version, and make the changes in ArangoDB so that everything still compiles.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: -
  - [ ] Backport for 3.9: -
  - [ ] Backport for 3.8: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Replication2/ReplicatedLog/NetworkMessages.cpp [hunk 1]
```diff
@@ -298,11 +298,10 @@ auto replicated_log::AppendEntriesRequest::fromVelocyPack(
   auto entries = std::invoke([&] {
     auto entriesVp = velocypack::ArrayIterator(slice.get("entries"));
     auto transientEntries = EntryContainer::transient_type{};
-    std::transform(
-        entriesVp.begin(), entriesVp.end(),
-        std::back_inserter(transientEntries), [](auto const& it) {
-          return InMemoryLogEntry(PersistingLogEntry::fromVelocyPack(it));
-        });
+    for (auto it : entriesVp) {
```
Comment (MBkkt) [inline arangod/Replication2/ReplicatedLog/NetworkMessages.cpp:301]: transientEntries.reserve

Comment (jsteemann) [reply arangod/Replication2/ReplicatedLog/NetworkMessages.cpp:301]: `reserve` function is not available here. This is not an std container, but an immer container.

File: arangod/Replication2/Methods.cpp [hunk 2]
```diff
@@ -699,9 +698,9 @@ struct ReplicatedLogMethodsCoordinator final
           std::vector<LogIndex> indexes;
           indexes.reserve(payloadSize);
           auto indexIter = velocypack::ArrayIterator(result.get("indexes"));
-          std::transform(
-              indexIter.begin(), indexIter.end(), std::back_inserter(indexes),
-              [](auto const& it) { return it.template extract<LogIndex>(); });
+          for (auto it : indexIter) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Replication2/Methods.cpp:701]: `reserve` function is not available here. This is not an std container, but an immer container.

File: tests/VPackDeserializer/test-types.h [hunk 3]
```diff
@@ -273,7 +271,7 @@ struct array_iterator {
   }
 
   bool operator!=(array_iterator const& other) const {
-    return iter.operator!=(other.iter);
+    return !iter.operator==(other.iter);
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply tests/VPackDeserializer/test-types.h:274]:
```
In file included from /home/jan/ArangoNext/tests/VPackDeserializer/BasicTests.cpp:26:
/home/jan/ArangoNext/tests/VPackDeserializer/test-types.h:274:17: error: no member named 'operator!=' in 'arangodb::velocypack::ArrayIterator'
    return iter.operator!=(other.iter);
           ~~~~ ^
```

[the remark this answers is not in the document]
Comment (MBkkt) [reply tests/VPackDeserializer/test-types.h:274]: Why don't call `iter != other.iter`?

[the remark this answers is not in the document]
Comment (jsteemann) [reply tests/VPackDeserializer/test-types.h:274]: I compiled with clang++-14.
For this compiler, ArrayIterator only has 
```
  [[nodiscard]] bool operator==(std::default_sentinel_t) const noexcept {
    return !valid();
  }
```
It don't think the compiler can auto-generate `operator!=(ArrayIterator const&)` for it

[the remark this answers is not in the document]
Comment (MBkkt) [reply tests/VPackDeserializer/test-types.h:274]: It *should*
https://godbolt.org/z/zEzKsrGro

**Review by MBkkt — APPROVED** [review]

File: arangod/RestHandler/RestAdminClusterHandler.cpp [hunk 4]
```diff
@@ -771,11 +772,13 @@ RestAdminClusterHandler::FutureVoid RestAdminClusterHandler::createMoveShard(
...
-  for (VPackArrayIterator i(shard); i != i.end(); i++) {
-    if (i.value().isEqualString(ctx->fromServer)) {
-      isLeader = i.isFirst();
+  velocypack::ArrayIterator it(shard);
+  while (it.valid()) {
+    if (it.value().isEqualString(ctx->fromServer)) {
+      isLeader = it.index() == 0;
       fromFound = true;
     }
+    it.next();
```
Comment (mpoeter) [inline arangod/RestHandler/RestAdminClusterHandler.cpp:781]: NIT
```suggestion
  for (velocypack::ArrayIterator it(shard); it.valid(); it.next()) {
    if (it.value().isEqualString(ctx->fromServer)) {
      isLeader = it.index() == 0;
      fromFound = true;
    }
```

File: tests/Aql/QueryHelper.cpp [hunk 5]
```diff
@@ -45,16 +45,18 @@ auto vpackHasNoneRecursive(VPackSlice slice) -> bool {
...
-    });
+    for (auto it : VPackArrayIterator(slice)) {
+      if (vpackHasNoneRecursive(it)) {
+        return true;
+      }
+    }
+  } else if (slice.isObject()) {
+    for (auto it :
+         VPackObjectIterator(slice, /*useSequentialIteration*/ true)) {
+      if (vpackHasNoneRecursive(it.value)) {
```
Comment (mpoeter) [inline tests/Aql/QueryHelper.cpp:56]: The old code uses `vpackHasNoneRecursive(pair.key) || vpackHasNoneRecursive(pair.value)`. Was this change intentional?

**Review by mpoeter — APPROVED** [review]
